codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -405,66 +439,68 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx, ChannelPro
-
-        frame.writeBytes(fragment);
-
-        // Write out the padding, if any.
-        frame.writeZero(padding);
-        return frame;
+        // TODO: same padding is applied to all frames, is this desired?
+        if (maxFragmentLength <= 0) {
+            return promise.setFailure(new IllegalArgumentException(
+                    "Padding [" + padding + "] is too large for max frame size [" + maxFrameSize + "]"));
+        }
+        SimpleChannelPromiseAggregator promiseAggregator =
+                new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+
+        if (headerBlock.isReadable()) {
+            // The frame header (and padding) only changes on the last frame, so allocate it once and re-use
+            final ByteBuf paddingBuf = padding > 0 ? ctx.alloc().buffer(padding).writeZero(padding) : null;
+            int fragmentReadableBytes = Math.min(headerBlock.readableBytes(), maxFragmentLength);
+            int payloadLength = fragmentReadableBytes + nonFragmentLength;
+            ByteBuf buf = ctx.alloc().buffer(CONTINUATION_FRAME_HEADER_LENGTH);
+            writeFrameHeaderInternal(buf, payloadLength, CONTINUATION, flags, streamId);
+            writePaddingLength(buf, padding);
+
+            do {
+                fragmentReadableBytes = Math.min(headerBlock.readableBytes(), maxFragmentLength);
+                ByteBuf fragment = headerBlock.readSlice(fragmentReadableBytes).retain();
+
+                payloadLength = fragmentReadableBytes + nonFragmentLength;
+                if (headerBlock.isReadable()) {
+                    ctx.write(buf.retain(), promiseAggregator.newPromise());
+                } else {
+                    // The frame header is different for the last frame, so re-allocate and release the old buffer
+                    flags = flags.endOfHeaders(true);
+                    buf.release();
+                    buf = ctx.alloc().buffer(CONTINUATION_FRAME_HEADER_LENGTH);
+                    writeFrameHeaderInternal(buf, payloadLength, CONTINUATION, flags, streamId);
+                    writePaddingLength(buf, padding);
+                    ctx.write(buf, promiseAggregator.newPromise());
+                }
+
+                ctx.write(fragment, promiseAggregator.newPromise());
+
+                // Write out the padding, if any.
+                if (paddingBuf != null) {
+                    ctx.write(paddingBuf.retain(), promiseAggregator.newPromise());
+                }
+            } while(headerBlock.isReadable());
+
+            if (paddingBuf != null) {
+                paddingBuf.release();
+            }
+        }
+        return promiseAggregator.doneAllocatingPromises();
     }
 
-    /**
-     * Writes the padding length field to the output buffer.
-     */
-    private static void writePaddingLength(int paddingLength, ByteBuf out) {
+    private static void writePaddingLength(ByteBuf buf, int paddingLength) {
         if (paddingLength > MAX_UNSIGNED_BYTE) {
```

Comment:
@nmittler - I reviewed the spec and I'm not sure padding is allowed to be > 8 bits.  https://tools.ietf.org/html/draft-ietf-httpbis-http2-16#section-6.1.  Even our `verifyPadding` method only allows a value up to 255.  I'm thinking this method should be changed to reflect that.  Am I missing something?